File: cpp/State.cpp
#include "State.h"
#include "Position.h"
#include "read_puzzle_file.h"
#include <array>

using namespace Sudoku;
using namespace std;

State::State() : empty_cell_count(N * N) {
  for (int i = 0; i < N; i++) {
    rows[i] = new Row(grid, i);
    columns[i] = new Column(grid, i);
  }
  for (int i = 0; i < B; i++) {
    for (int j = 0; j < B; j++) {
      blocks[i][j] = new Block(grid, i, j);
    }
  }
}

State::State(const State& source)
    : empty_cell_count(source.empty_cell_count), grid(source.grid) {
  for (int i = 0; i < N; i++) {
    rows[i] = new Row(grid, *source.rows[i]);
    columns[i] = new Column(grid, *source.columns[i]);
  }
  for (int i = 0; i < B; i++) {
    for (int j = 0; j < B; j++) {
      blocks[i][j] = new Block(grid, *source.blocks[i][j]);
    }
  }
}

State::~State() {
  for (int i = 0; i < N; i++) {
    delete rows[i];
    delete columns[i];
  }
  for (int i = 0; i < B; i++) {
    for (int j = 0; j < B; j++) {
      delete blocks[i][j];
    }
  }
}

void State::load_from_file(const std::string& filename) {
  int raw_grid[N][N];
  read_puzzle_file(filename, raw_grid);

  for (int row_index = 0; row_index < N; row_index++) {
    for (int column_index = 0; column_index < N; column_index++) {
      cell_t value = raw_grid[row_index][column_index];
      if (value) {
        place(value, Position(row_index, column_index));
      }
    }
  }
}

void State::place(cell_t value, const Position& position) {
  grid(position).place(value);

  for (Section* section : sections_for_position(position)) {
    section->place(value);
  }

  empty_cell_count--;
}

const array<Section*, State::SECTIONS_PER_POSITION>
State::sections_for_position(const Position& position) {
  return array<Section*, SECTIONS_PER_POSITION>{
      &row(position),
      &column(position),
      &block(position),
  };
}

void State::each_section(SectionFunction operate) {
  for (int i = 0; i < N; i++) {
    operate(*rows[i]);
  }

  for (int i = 0; i < N; i++) {
    operate(*columns[i]);
  }

  for (int i = 0; i < B; i++) {
    for (int j = 0; j < B; j++) {
      operate(*blocks[i][j]);
    }
  }
}

File: ts/src/State.ts
import fs from 'fs-extra';
import _ from 'lodash';
import { B, N } from './core';
import Grid from './Grid';
import GridPosition from './GridPosition';
import { Block, Column, Row, Section } from './Sections';
import Cell from './Cell';

export default class State {
  grid: Grid;
  emptyCellCount: number;

  rows: Row[];
  columns: Column[];
  blocks: Block[][];

  constructor(source?: State) {
    if (source) {
      this.grid = source.grid.clone();
      this.emptyCellCount = source.emptyCellCount;

      this.rows = source.rows.map(row => row.clone(this.grid));
      this.columns = source.columns.map(column => column.clone(this.grid));
      this.blocks = source.blocks.map(blockRow =>
        blockRow.map(block => block.clone(this.grid))
      );
    } else {
      this.grid = new Grid();
      this.emptyCellCount = N * N;

      this.rows = _.times(N, i => new Row(this.grid, i));
      this.columns = _.times(N, i => new Column(this.grid, i));
      this.blocks = _.times(B, br =>
        _.times(B, bc => new Block(this.grid, br, bc))
      );
    }
  }

  clone() {
    return new State(this);
  }

  public loadFromFile(filename: string) {
    const rawTable = _.chain(fs.readFileSync(filename, 'utf-8'))
      .split('\n')
      .map(line => line.split('').map(toCellValue))
      .value();

    this.grid.eachPosition(position => {
      const value = rawTable[position.row][position.column];
      if (value) {
        this.place(value, position);
      }
    });
  }

  public eachCell(fn: (c: Cell, p: GridPosition) => any) {
    this.grid.eachPosition(position => fn(this.grid.cell(position), position));
  }

  public eachSection(fn: (s: Section) => any) {
    this.eachRow(fn);
    this.eachColumn(fn);
    this.eachBlock(fn);
  }

  public eachRow(fn: (r: Row) => any) {
    this.rows.forEach(fn);
  }

  public eachColumn(fn: (c: Column) => any) {
    this.columns.forEach(fn);
  }

  public eachBlock(fn: (b: Block) => any) {
    this.blocks.forEach(blockRow => blockRow.forEach(fn));
  }

  public place(value: number, position: GridPosition) {
    this.grid.cell(position).place(value);

    this.sectionsForPosition(position).forEach(section => section.place(value));

    this.emptyCellCount--;
  }

  sectionsForPosition = (position: GridPosition) => [
    this.row(position),
    this.column(position),
    this.block(position),
  ];

  row = (position: GridPosition) => this.rows[position.row];
  column = (position: GridPosition) => this.columns[position.column];
  block = (position: GridPosition) =>
    this.blocks[blockIndex(position.row)][blockIndex(position.column)];
}

function blockIndex(index: number) {
  return Math.floor(index / B);
}

function toCellValue(char: string) {
  return char.match(/\s+/) ? null : _.toNumber(char);
}
